### PR TITLE
Add allow_partial_search_results flag to search requests with default setting true

### LINKED ISSUE #27435: New query flag: allow_partial_results with default set to false.
There are various reasons today why a search could fail to produce all of the required results:

1) A failure to execute within a `timeout` associated with the search
2) Missing shards (ie a red index)
3) Failed shards - e.g. an error parsing the query

Until now elasticsearch has  attempted to to return partial results under these circumstances and report the partial status in flags in the search response e.g. `timed_out:true`. 
Sadly, today we expect most client applications will pay little attention to these flags and present partial results to clients with no indication that the results are incomplete.

This issue proposes we fix this by providing a new cluster level setting called `default_allow_partial_results` flag which is set to false and will throw an error to the client application in the event of any of the above 3 conditions. This can be overridden in individual search requests using a `allow_partial_results` flag.

### PR DESCRIPTION
**Pull request by markharwood**
When false, will error if search either timeouts, has partial errors or has missing shards rather
than returning partial search results. A cluster-level setting provides a default for searches
with no flag.

Closes #27435

### CODE REVIEW

**Comment by markharwood** [comment]
Initial commit here just for CI checking that forward-port of 6.x-based https://github.com/elastic/elasticsearch/pull/27906 was successful. Once green I'll add a commit that changes the default for `allowPartialSearchResults` to false (meaning errors will be thrown by default when it looks like partial results would be returned)

File: rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/190_percentiles_hdr_metric.yml [hunk 1]
```diff
@@ -418,6 +418,7 @@ setup:
 
   - do:
       search:
+        allow_partial_search_results: true
```
Comment (colings86) [inline rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/190_percentiles_hdr_metric.yml]: Could you explain why this is needed? This test doesn't appear to test partial results?

File: rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/13_fields.yml [hunk 2]
```diff
@@ -71,6 +71,7 @@ setup:
 
   - do:
       search:
+        allow_partial_search_results: true
```
Comment (colings86) [inline rest-api-spec/src/main/resources/rest-api-spec/test/indices.stats/13_fields.yml]: Could you explain why this is needed? This test doesn't appear to test partial results?

**Review by colings86 — COMMENTED** [review]
I left a couple of comments int eh tests.

File: rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/190_percentiles_hdr_metric.yml (hunk 1 above)
Comment (markharwood) [reply rest-api-spec/src/main/resources/rest-api-spec/test/search.aggregation/190_percentiles_hdr_metric.yml]: That was a necessity from a previously failing trial on 6.x - if I recall correctly it was for searches that rely on the use of field data. Ditto the other test

**Comment by markharwood** [comment]
Reverted the code that defaults searches to strict (error on partial results). Will open a separate issue once this is pushed to consider changing the default to strict.

**Comment by jasontedor** [comment]
If I can assist in integrating a change like this with cross-cutting BWC implications, please let me know. I am always happy to help.

**Comment by markharwood** [comment]
@jasontedor Thanks for the offer of help. I have 2 PRs lined up ready to push (if I can get an OK on this one, @colings86 ?)
This PR is for master and is failing on BWC, but I have a seperate [6.x PR](https://github.com/elastic/elasticsearch/pull/27906) which is passing tests.
I thought the best way to sequence this might be to: 
1) push the 6.x PR which should be green but will then cause all master BWC search tests to fail (master will not de/serialize search requests to 6.x)
2) push this master PR which should then hopefully be green

Not sure if there needs to be a gap between 1) and 2) waiting for the 6.x build to make it through the  intake tests?

After all this is done I'll open a separate issue to open the debate on whether new cluster default setting is to fail on partial results.

**Review by colings86 — APPROVED** [review]
LGTM

**Comment by jasontedor** [comment]
@markharwood I think that this would help keep the build green every step of the way and let you integrate this at a leisurely pace:
 - push a commit to this PR switching all version-dependent serialization logic to `Version.V_7_0_0`
 - push a commit to this PR that disables BWC testing; make a change to the root `build.gradle` setting `ext.bwc_tests_enabled` to `false` in [all projects](https://github.com/elastic/elasticsearch/blob/eea9ee57ddf9615aed2edc8440b11228f5fbb7af/build.gradle#L148).
 - at this point, you should be able to get a completely green build
 - merge this PR
 - prepare a backport of this change to 6.x, changing the version-dependent serialization logic to `Version.V_6_3_0`, and setting `ext.bwc_tests_enabled` to `true`
 - at this point, you should be able to get a completely green build on 6.x, and integrating this change there will have no impact on master
 - backport this PR with the above changes to 6.x
 - prepare a commit against master that changes the version-dependent serialization logic to `Version.V_6_3_0` and reenables `ext.bwc_tests_enabled`; and this point you can run the BWC tests and they should be green
 - integrate the above change of the version logic and reenabling BWC into master

**Pull request merged**